### ISSUE TITLE
Inconsistencies between Readme, Code and paper

### ISSUE THREAD

**Issue by jgcb00** [TianjinYellow/StableSPAM#3]
Hi,
Thanks for your work, 

I'm a bit confuse in the Readme it says :
```
optimizer = StableSPAM(model.parameters(), lr=0.001,gamma1=0.7,gamma2=0.9,gamma3=0.999,total_T=20000)
```
but the init function of StableSPAM doesn't have any gamma3 so I suppos it's equivalent to theta ?
```
class StableSPAM(Optimizer):
    def __init__(self, params, lr=1e-3, betas=(0.9, 0.999), eps=1e-8,
                 weight_decay=0, amsgrad=False,gamma1=0.7,gamma2=0.9,theta=0.999,total_T=None,eta_min=0.5,update_proj_gap=1000):
```

In the training script for bf16
The update proj gap is set to 500 but in the paper it's mentioned 1000

I tried to implement it in nanoGPT but after the first optimizations, it gives me a loss of `nan` where SPAM for example works out of the box,
I changed the parameters to make it works with BF16 : 
```
class StableSPAM(Optimizer):
    def __init__(self, params, lr=8e-4, betas=(0.9, 0.999), eps=1e-8,
                 weight_decay=0, amsgrad=False,gamma1=0.85,gamma2=0.99999,theta=0.999,total_T=None,eta_min=0.5,update_proj_gap=1000):
 ```
I'm training a 170m GPT-2 

Any ideas ? 
Regards

**Comment by TianjinYellow** [comment]
Hi,

Sorry for the confusion.
You are right—there are mistakes in the training scripts for BF16 and the StableSPAM hyper-parameters. I will update them soon.

When using the following hyper-parameters, did you still encounter NaN loss for training 170m GPT-2?
class StableSPAM(Optimizer):
    def __init__(self, params, lr=8e-4, betas=(0.9, 0.999), eps=1e-8,
                 weight_decay=0, amsgrad=False,gamma1=0.85,gamma2=0.99999,theta=0.999,total_T=None,eta_min=0.5,update_proj_gap=1000):

Best regards,

**Comment by jgcb00** [comment]
Hi,
Indeed I encountered NaN with these hyper-parameters, but I found the cause, here the fix I applied : 
```
# FIX: Avoid division by zero
if grad_norm > 0:
    grad = grad / grad_norm * c_norm_t
# If grad_norm is zero, keep grad as zeros (no need to normalize)
```

**Comment by TianjinYellow** [comment]
Hi, 

I see. There is a gradient vanishing in your case. Thank you for pointing out this. I will integrate this condition into StableSPAM soon.

Thanks once again.

Best regards,
